### DISCUSSION TITLE
Dynamic Link, useRouter and usePathname locales

### DISCUSSION THREAD

**Discussion by awinogrodzki** [amannn/next-intl#597] [Q&A]
### Is your feature request related to a problem? Please describe.

I have recently upgraded to `3.0.0-rc.6`. There is one issue I am having in regards to this upgrade.

```tsx
import {
  createLocalizedPathnamesNavigation,
  Pathnames
} from 'next-intl/navigation';
 
export const locales = ['en', 'de'] as const;
 
// The `pathnames` object holds pairs of internal
// and external paths, separated by locale.
export const pathnames = {
  // If all locales use the same pathname, a
  // single external path can be provided.
  '/': '/',
  '/blog': '/blog',
 
  // If locales use different paths, you can
  // specify each external path per locale.
  '/about': {
    en: '/about',
    de: '/ueber-uns'
  },
 
  // Dynamic params are supported via square brackets
  '/news/[articleSlug]-[articleId]': {
    en: '/news/[articleSlug]-[articleId]',
    de: '/neuigkeiten/[articleSlug]-[articleId]'
  },
 
  // Also (optional) catch-all segments are supported
  '/categories/[...slug]': {
    en: '/categories/[...slug]',
    de: '/kategorien/[...slug]'
  }
} satisfies Pathnames<typeof locales>;
 
export const {Link, redirect, usePathname, useRouter, getPathname} =
  createLocalizedPathnamesNavigation({locales, pathnames});
```

My app features locales dynamic per user. User can configure the set of languages the app supports. 

`createLocalizedPathnamesNavigation` requires `locales` to be known at build time, the same for the whole app.

### Describe the solution you'd like

Ideally, I would use internal `next/navigation` library features instead of decorated ones 

### Describe alternatives you've considered

I am still trying to figure out how I can create those components/hooks dynamically and it's a pain. Probably I would have to share them using React Context, which is quite painful.

**Comment by awinogrodzki** [comment]
Here's how I construct intl middleware:

```tsx
export const handleI18nRouting = (client?: Client) =>
  createIntlMiddleware({
    locales:
      client?.settings && isFeatureEnabled(client.settings, Feature.Translation)
        ? getFeatureOption(client.settings, Feature.Translation, 'locales', locales)
        : locales,
    defaultLocale:
      client?.settings && isFeatureEnabled(client.settings, Feature.Translation)
        ? getFeatureOption(client.settings, Feature.Translation, 'defaultLocale', defaultLocale)
        : defaultLocale,
    localeDetection: client?.settings
      ? isFeatureEnabled(client.settings, Feature.Translation)
      : false,
  });
```

**Comment by awinogrodzki** [comment]
Here's what I've come up so far with `3.0.0-rc6`
```tsx
type LocalizedNavigation = ReturnType<typeof createLocalizedPathnamesNavigation<any, any>>;

interface PageRouterContextValue {
  navigation: LocalizedNavigation;
}

const PageRouterContext = createContext<PageRouterContextValue>({
  get navigation(): LocalizedNavigation {
    throw new Error('Context used outside provider');
  },
});

interface PageRouterProviderProps {
  children: React.ReactNode;
}

export function PageRouterProvider({ children }: PageRouterProviderProps) {
  const { settings, isFeatureEnabled, getFeatureOption } = useSettings();
  const navigation = useMemo(
    () =>
      createLocalizedPathnamesNavigation({
        locales: isFeatureEnabled(Feature.Translation)
          ? getFeatureOption(Feature.Translation, 'locales', locales)
          : locales,
        pathnames: {},
      }),
    [settings]
  );
  return (
    <PageRouterContext.Provider value={{ navigation: navigation as unknown as LocalizedNavigation }}>
      {children}
    </PageRouterContext.Provider>
  );
}
```

**Comment by aXenDeveloper** [comment]
You can import `json` file into `createLocalizedPathnamesNavigation`.
Something like this:

```tsx
import configs from '~/config.json';

import { getRequestConfig } from 'next-intl/server';
import { createSharedPathnamesNavigation } from 'next-intl/navigation';

export default getRequestConfig(async ({ locale }) => {
  const messagesFormApps = await Promise.all(
    configs.applications.map(async app => {
      return {
        ...(await import(`~/langs/${locale}/${app}.json`)).default
      };
    })
  );

  return {
    messages: {
      ...messagesFormApps.reduce((acc, messages) => ({ ...acc, ...messages }), {})
    },
    timeZone: 'UTC'
  };
});

export const { Link, redirect, usePathname, useRouter } = createSharedPathnamesNavigation({
  locales: configs.languages.locales.filter(locale => locale.enabled).map(locale => locale.key)
});

```

**Comment by amannn** [comment]
`createLocalizedPathnamesNavigation` is really about providing static type safety for the pathnames that are available to your application. Providing the resulting APIs via context doesn't seem right to me.

How many locales does your app handle? Any chance you can create the navigation APIs in the outer module scope with all supported locales and restrict them within the app to the subset that the user supports?

I'll move this to a discussion since this seems to be more of a usage question.

**Reply by awinogrodzki** [reply]
Well, I have a list of supported languages that user can pick from, and it's this:

```tsx
export const LOCALE_LIST = [
  'af',
  'ak',
  'am',
  'an',
  'ar',
  'as',
  'ay',
  'az',
  'be',
  'bg',
  'bm',
  'bn',
  'br',
  'bs',
  'ca',
  'co',
  'cs',
  'cy',
  'da',
  'de',
  'dv',
  'ee',
  'el',
  'en',
  'eo',
  'es',
  'et',
  'eu',
  'fa',
  'fi',
  'fo',
  'fr',
  'fy',
  'ga',
  'gd',
  'gl',
  'gn',
  'gu',
  'ha',
  'he',
  'hi',
  'hr',
  'ht',
  'hu',
  'hy',
  'ia',
  'id',
  'ig',
  'is',
  'it',
  'ja',
  'jv',
  'ka',
  'kk',
  'km',
  'kn',
  'ko',
  'ku',
  'ky',
  'la',
  'lb',
  'lg',
  'ln',
  'lo',
  'lt',
  'lv',
  'mg',
  'mi',
  'mk',
  'ml',
  'mn',
  'mr',
  'ms',
  'mt',
  'my',
  'nb',
  'ne',
  'nl',
  'nn',
  'no',
  'ny',
  'oc',
  'om',
  'or',
  'pa',
  'pl',
  'ps',
  'pt',
  'qu',
  'rm',
  'ro',
  'ru',
  'rw',
  'sa',
  'sd',
  'si',
  'sk',
  'sl',
  'sm',
  'sn',
  'so',
  'sq',
  'sr',
  'st',
  'su',
  'sv',
  'sw',
  'ta',
  'te',
  'tg',
  'th',
  'ti',
  'tk',
  'tl',
  'tn',
  'to',
  'tr',
  'ts',
  'tt',
  'tw',
  'ug',
  'uk',
  'ur',
  'uz',
  'vi',
  'wa',
  'wo',
  'xh',
  'yi',
  'yo',
  'zh',
  'zu',
];
```

Each user can create their own website:

1. `User A` creates `user-a.com`, sets `en` default language and `de`, `fr` as their supported languages
2. `User B` creates `user-b.fr`, sets `fr` as default language and `en` as supported language

When choosing which language their website will support, they have a predefined list of locales mentioned in `LOCALE_LIST`.

I guess this should also work, wouldn't it?:

```tsx
export const { Link, redirect, usePathname, useRouter } = createSharedPathnamesNavigation({
  locales: LOCALE_LIST
});
```

**Comment by markomitranic** [comment]
Just a bump on this one :) - I have a similar use case with a lot of my clients (e-commerce), where the entirety of the configuration sits inside a remote CMS. They can go into the admin panel, any time of day, and create a new market, or assign one or multiple languages to a market, as well as a default language, localePrefix strategy and a domain for each market.

What we currently do is basically rebuild the app on-demand when a change is needed, which thankfully isn't that often.

**Reply by amannn** [reply]
Yeah, I guess there's a fair point that locales can't always be known statically. Making the `locales` argument optional for the navigation APIs should work, right? Note you'd still have to pass the argument to the middleware though.

I've created https://github.com/amannn/next-intl/pull/784 to address this, let me know if this works for you!

/cc @awinogrodzki
